### ISSUE TITLE
union within nested list leads to incorrect error

### ISSUE THREAD

**Issue by jakevdp** [python/mypy#16192]
**Bug Report**

When a parameter is annotated as a union within nested lists, nested lists of one union member are incorrectly flagged as a mismatched type.

**To Reproduce**

Contents of `repro.py`:
```python
def f(x: list[list[int | float]]) -> None:
    return None
x = [[1]]
f(x)
```
```
$ python --version
Python 3.10.12

$ python -m mypy --version
mypy 1.5.1 (compiled: yes)

$ python -m mypy repro.py
repo.py:4: error: Argument 1 to "f" has incompatible type "list[list[int]]"; expected "list[list[int | float]]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

**Expected Behavior**
I expect no error, because `list[list[int]]` should be treated as compatible with `list[list[int | float]]`.

**Environment Info**

- Mypy version used: 1.5.1
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none (above execution is in a clean evironment)
- Python version used: 3.10.12

**Comment by JelleZijlstra** [comment]
https://mypy.readthedocs.io/en/stable/common_issues.html#invariance-vs-covariance
